transformer_engine/paddle/layer/attention.py

```diff
@@ -0,0 +1,458 @@
+                                               aux_ctx_tensors, ctx.max_seqlen, ctx.qkv_dtype,
+                                               ctx.attn_scale, ctx.dropout_p, ctx.set_zero,
+                                               ctx.qkv_layout, ctx.attn_bias_type,
+                                               ctx.attn_mask_type)
+
+        # if no_bias, return dqkv
+        if ctx.attn_bias_type == "no_bias":
+            return (dqkv, None, None)
+        # else, return (dqkv, dbias)
+        return (dqkv, None, rest[0], None)
+
+
+class FusedAttnFuncPackedKV(paddle.autograd.PyLayer):
+    """Function for FusedAttention with packed KV input"""
+
+    @staticmethod
+    def forward(ctx, q, kv, cu_seqlens_q, cu_seqlens_kv, attn_bias, rng_state, max_seqlen_q,
+                max_seqlen_kv, attn_scale, qkv_dtype, dropout_p, set_zero, qkv_layout,
+                attn_bias_type, attn_mask_type, is_training):
+        out, aux_ctx_tensors = fused_attn_fwd_kvpacked(q, kv, cu_seqlens_q, cu_seqlens_kv,
+                                                       rng_state, is_training, max_seqlen_q,
+                                                       max_seqlen_kv, qkv_dtype, attn_bias,
+                                                       attn_scale, dropout_p, set_zero, qkv_layout,
+                                                       attn_bias_type, attn_mask_type)
+
+        ctx.save_for_backward(q, kv, out, cu_seqlens_q, cu_seqlens_kv, rng_state, aux_ctx_tensors)
+        ctx.max_seqlen_q = max_seqlen_q
+        ctx.max_seqlen_kv = max_seqlen_kv
+        ctx.qkv_dtype = qkv_dtype
+        ctx.attn_scale = attn_scale
+        ctx.dropout_p = dropout_p
+        ctx.set_zero = set_zero
+        ctx.qkv_layout = qkv_layout
+        ctx.attn_bias_type = attn_bias_type
+        ctx.attn_mask_type = attn_mask_type
+
+        return out
+
+    @staticmethod
+    def backward(ctx, d_out):
+        q, kv, out, cu_seqlens_q, cu_seqlens_kv, rng_state, aux_ctx_tensors = ctx.saved_tensor()
+        dq, dkv, *rest = fused_attn_bwd_kvpacked(q, kv, cu_seqlens_q, cu_seqlens_kv, rng_state, out,
+                                                 d_out, aux_ctx_tensors, ctx.max_seqlen_q,
+                                                 ctx.max_seqlen_kv, ctx.qkv_dtype, ctx.attn_scale,
+                                                 ctx.dropout_p, ctx.set_zero, ctx.qkv_layout,
+                                                 ctx.attn_bias_type, ctx.attn_mask_type)
+
+        # if no_bias, return dq, dkv
+        if ctx.attn_bias_type == "no_bias":
+            return (dq, dkv, None, None, None)
+        # else, return (dq, dkv, dbias)
+        return (dq, dkv, None, None, rest[0], None)
+
+
+class DotProductAttention(paddle.nn.Layer):
+    """Dot Product Attention Layer
+    BMM1 -> softmax + dropout -> BMM2
+    """
+
+    def __init__(self,
+                 norm_factor: int,
+                 attention_dropout: float = 0.0,
```

Comment:
Discussion; should the dropout default value be `0.0`?

Reply:
The `TransformerLayer` API has `0.1` default so we should stick to that.

Reply:
Additionally, since `DotProductAttention` and `MultiHeadAttention` are APIs, we should improve those docs (hopefully should just be copy and pasting a bunch of stuff).

Reply:
Done. I've set the default value of `dropout` to `0.1` and added docs to `DotProductAttention` and `MultiHeadAttention`.